### PR TITLE
dotnet new template - Support unattended

### PR DESCRIPTION
**Pull request by warrenbuckley**
## Updates to .NETCore dotnet new templates

Updated logo image to be transparent

Renames Umbraco Solution to Umbraco Project as it does not create a SLN

Adds support for doing an auto Nuget Restore with an optional switch
https://github.com/dotnet/templating/wiki/Post-Action-Registry#restore-nuget-packages

Auto generates HTTP & HTTPS port numbers that are available to use
https://github.com/dotnet/templating/wiki/Available-Symbols-Generators#port

Support for Unattended Installs with Name, Email & Password & Connection String as CLI params & in VS
https://github.com/dotnet/templating/wiki/Reference-for-comment-syntax#json-files
https://github.com/dotnet/templating/wiki/Reference-for-template.json#computed-symbol

### Create Project
`dotnet new umbraco -n MyNewProject`

### Create Project using SQLCE (Windows only)
`dotnet new umbraco -n MyNewProjectWithCE --UseSqlCe`

### Create Project & skip auto Nuget Restore
`dotnet new umbraco -n MyNewProject --SkipRestore`

### Create Project for Unattended Install (Skip UI Installer)
`dotnet new umbraco -n MyNewProject --FriendlyName "Friendly User" --Email warren@umbraco.com --Password password1234 --ConnectionString "Server=(localdb)\\Umbraco;Database=MyDatabase;Integrated Security=true"`

## CLI
image

```
Umbraco Project (C#)
Author: Umbraco HQ
Description: An empty Umbraco Project ready to get started
Options:
  -v|--version              The version of Umbraco to load using NuGet
                            string - Optional
                            Default: 9.0.0-beta001

  -P|--PackageTestSiteName  The name of the package this should be a test site for (Default: '')
                            text - Optional

  -U|--UseSqlCe             Adds the required dependencies to use SqlCE (Windows only) (Default: false)
                            bool - Optional
                            Default: false

  -F|--Framework            The target framework for the project
                                net5.0    - Target net5.0
                                net6.0    - Target net6.0
                            Default: net5.0

  -S|--SkipRestore          If specified, skips the automatic restore of the project on create
                            bool - Optional
                            Default: false

  -Fr|--FriendlyName        The friendly name of the user for Umbraco login when using Unattended install (Without installer wizard UI)
                            text - Optional

  -E|--Email                Email to use for Umbraco login when using Unattended install (Without installer wizard UI)
                            text - Optional

  -Pa|--Password            Password to use for Umbraco login when using Unattended install (Without installer wizard UI)
                            text - Optional

  -C|--ConnectionString     Database connection string when using Unattended install (Without installer wizard UI)
                            text - Optional
```

## Visual Studio
image

For information on how to enable dotnet new templates to appear in Visual Studio
https://devblogs.microsoft.com/dotnet/net-cli-templates-in-visual-studio/

image

## Testing

`dotnet new --install C:\Code\Umbraco-CMS_netcore\build.out\Umbraco.Templates.9.0.0-beta001.nupkg`

`dotnet nuget add source "https://www.myget.org/F/umbraconightly/api/v3/index.json" -n "Umbraco Nightlies"`

` dotnet new umbraco -n MyNewProject3 --FriendlyName "Friendly User" --Email warren@umbraco.com --Password password1234 --ConnectionString "Server=(localdb)\\Umbraco;Database=MyDatabase;Integrated Security=true" --version 9.0.0-beta001.20210421.2`

### CODE REVIEW

File: build/templates/UmbracoProject/.template.config/template.json [hunk 1]
```diff
@@ -67,6 +77,32 @@
...
+            "defaultValue": "false"
+        },
+        "HttpPort": {
+            "type": "generated",
+            "generator": "port",
+            "replaces": "HTTP_PORT_FROM_TEMPLATE",
+            "parameters": {
+                "high": 65535,
+                "low": 1024,
+                "fallback": 9000
```
Comment (bergmania) [inline build/templates/UmbracoProject/.template.config/template.json:94]:
```suggestion
                "fallback": 5000
```

I think fallback should be 5000 as the default value from ASP.NET Core projects.. 9000 looks like our internal ports

File: build/templates/UmbracoProject/.template.config/template.json [hunk 2]
```diff
@@ -67,6 +77,32 @@
...
+            }
+        },
+        "HttpsPort": {
+            "type": "generated",
+            "generator": "port",
+            "replaces": "HTTPS_PORT_FROM_TEMPLATE",
+            "parameters": {
+                "high": 65535,
+                "low": 1024,
+                "fallback": 9001
```
Comment (bergmania) [inline build/templates/UmbracoProject/.template.config/template.json:104]:
```suggestion
                "fallback": 5001
```

I think fallback should be 5001 as the default value from ASP.NET Core projects.. 9001 looks like our internal ports

**Review by elit0451 — COMMENTED** [review]
Overall it looks very nice! Left some inline comments here and there.

I had problems with **Unattended Install** - I couldn't login into the back-office. My steps were the following:
- Create a new DB in Localdb;
- Executed the command you gave as an example for *Unattended Install*;
- run `dotnet run`
  - On this step, I could see that the DB tables are being created but after that, the execution stopped, so I needed to run the command once again - only after that I got the info for the address and port where Umbraco is running.
- Typed in the same username and pass as the ones used in the dotnet new command.
- Got incorrect login    

I can confirm that the connection string was added to the appsettings.Development.json and that the DB was populated but not sure why I can't login...

**Comment by bergmania** [comment]
> I can confirm that the connection string was added to the appsettings.Development.json and that the DB was populated but not sure why I can't login...

Could be related to another bug. Try again if you merge in latest netcore/dev, then it should work

**Comment by elit0451** [comment]
Still the same experience, unfortunately - execution of `dotnet run` stops after the table are created and after I re-run it, can't login into the back-office. Plus after the merge the Programs.cs and Startup file have `Umbraco.Cms.Web.UI.NetCore` as namespace

@warrenbuckley can you have a look?

File: build/build.ps1 [hunk 3]
```diff
@@ -344,13 +344,11 @@
     $this.CopyFiles("$templates", "*", "$tmp\Templates")
 
     Write-Host "Copy files for dotnet templates"
-    $this.CopyFiles("$src\Umbraco.Web.UI.NetCore", "Program.cs", "$tmp\Templates\UmbracoSolution")
-    $this.CopyFiles("$src\Umbraco.Web.UI.NetCore", "Startup.cs", "$tmp\Templates\UmbracoSolution")
-    $this.CopyFiles("$src\Umbraco.Web.UI.NetCore", "appsettings.json", "$tmp\Templates\UmbracoSolution")
-    $this.CopyFiles("$src\Umbraco.Web.UI.NetCore", "appsettings.Development.json", "$tmp\Templates\UmbracoSolution")
```
[the remark this answers is not in the document]
Comment (warrenbuckley) [reply build/build.ps1:350]: @elit0451 the reason is to do with having the template dynamically change & add extra settings to the appsettings.json and these extra things don't make sense being shipped in the main CMS product.

https://github.com/umbraco/Umbraco-CMS/pull/10107/files#diff-c405c7dafcf6153a9dac1de7a41fd317393b0a37ae68ab54f0490b7ebf42b456R19

So yes unfortunately we need to maintain two locations of these

File: build/templates/UmbracoProject/appsettings.Development.json [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          ]
+        }
+      }
+    ]
+  },
+  //#if (UsingUnattenedInstall)
+  "ConnectionStrings": {
+    "umbracoDbDSN": "CONNECTION_FROM_TEMPLATE"
+  },
+  //#endif
```
Comment (warrenbuckley) [inline build/templates/UmbracoProject/appsettings.Development.json:23]: Here is a section in the new appsettings I mentioned that will add in this extra setting if they have defined a value @elit0451

**Comment by warrenbuckley** [comment]
This worked fine for me @elit0451

I uninstalled the templates

Done a build with the powershell script

Then installed them from the build.out folder
`dotnet new --install C:\Code\Umbraco-CMS_netcore\build.out\Umbraco.Templates.9.0.0-beta001.nupkg`

Added the dotnet Nuget source to the MyGet feed of nightlies
`dotnet nuget add source "https://www.myget.org/F/umbraconightly/api/v3/index.json" -n "Umbraco Nightlies"`

Then run the template
`dotnet new umbraco -n MyNewProject3 --FriendlyName "Friendly User" --Email warren@umbraco.com --Password password1234 --ConnectionString "Server=(localdb)\\Umbraco;Database=MyDatabase;Integrated Security=true" --version 9.0.0-beta001.20210421.2`

Ran the site with `dotnet run` and found the site URL in the console output and clicked it and logged into the backoffice
image

File: build/build.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (elit0451) [reply build/build.ps1:350]: Makes sense then 👍

**Comment by elit0451** [comment]
Yep, you are right! 😄 today it worked for me as it should. 
Great work, merging now 🎊

**Pull request merged**